personal/privacyidea/tokens/class_MFARegistrationToken.inc

```diff
@@ -112,29 +121,42 @@ class MFARegistrationToken extends MFATokenHelper implements MFAToken
         //       and it should gather the tokens from endpoint /token/ itself.
         $this->utils->getTokensOfUser($this->mfaAccount->getUid(), $this->mfaAccount->userRealm);
 
-        $mfaRecoveryKey = $requestAnswer["detail"]["registrationcode"];
+        $this->mfaRecoveryKey = $requestAnswer["detail"]["registrationcode"];
+
+        $this->smarty->assign("tokenDescription", $this->tokenDescription);
+        $this->smarty->assign("mfaRecoveryKey", $this->mfaRecoveryKey);
+
+        return true;
+    }
+
+    /** @return bool */
+    protected function preparePhase2()
+    {
+        if (!isset($this->mfaRecoveryKey) || empty($this->mfaRecoveryKey)) {
+            return false;
+        }
 
         require_once(dirname(__FILE__) . "/../PDF/class_RecoveryKeyPDF.inc");
 
+        /* Get givenName and sn via UID */
+        $ldap = $this->config->get_ldap_link();
+        $ldap->cd($this->config->current['BASE']);
+        $uid = $this->mfaAccount->getUid();
+        $ldap->search("(uid=$uid)", array("givenName", "sn"));
+        $attributes = $ldap->fetch();
+
         $infoPage = new RecoveryKeyPDF();
         $infoPage->RecoveryKeyPDF(
-            $this->mfaAccount->givenName,
-            $this->mfaAccount->sn,
+            $attributes["givenName"][0],
+            $attributes["sn"][0],
             $this->mfaAccount->getUid(),
             $mfaRecoveryKey
         );
 
+        ob_end_clean(); // Clean (erase) the output buffer and turn off output buffering.
+        // TODO: Token description could be neat for pdfFileName too.
         $pdfFileName = sprintf(_("recovery_key_%s.pdf"), $this->mfaAccount->getUid());
-        $output = $infoPage->output('S', $pdfFileName, true);
-
-        $this->smarty->assign("tokenDescription", $this->tokenDescription);
-        $this->smarty->assign("mfaRecoveryKeyPDF_BASE64", base64_encode($output));
-        $this->smarty->assign("mfaRecoveryKeyPDF_fileName", $pdfFileName);
-        $this->smarty->assign("mfaRecoveryKey", $mfaRecoveryKey);
-
-        // DEBUG DEBUG DEBUG
-        // $pdfFilePath = sys_get_temp_dir() . PATH_SEPARATOR . $pdfFileName;
-        // file_put_contents($pdfFilePath, $output);
+        $infoPage->output('D', $pdfFileName, true);
```

Comment:
I'm currently not able to test this, does this imply `Content-Disposition: attachment`, i.e. is the file directly downloaded rather than displayed in the browser's PDF viewer? (If not his should be set.)

Reply:
I don't know about `Content-Disposition: attachment` but the file will be directly downloaded (and opened via file:/// URI on atleast Firefox Debian 12).

Reply:
In chromium, the file downloads normally without opening it.